File: core/runtime/service_lifecycle_engine.py
from __future__ import annotations

from typing import Any, Dict, List


LIFECYCLE_PHASES = ("bootstrap", "ready", "serving", "draining", "stopped")


def infer_service_lifecycle(
    services: List[Dict[str, Any]],
    parser_evidence: List[str],
) -> Dict[str, Any]:
    phases = []
    for svc in sorted(services, key=lambda s: str(s.get("name", ""))):
        phase = str(svc.get("phase", "bootstrap")).lower()
        if phase not in LIFECYCLE_PHASES:
            phase = "bootstrap"
        phases.append({"name": svc.get("name"), "phase": phase})
    return {
        "services": phases,
        "evidence": sorted(set(parser_evidence)),
        "grounded": bool(parser_evidence),
        "deterministic": True,
    }

File: src/runtime/serviceLifecycleEngine.ts
/**
 * Converted from Python: core/runtime/service_lifecycle_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "./pyCompat.js";

export let LIFECYCLE_PHASES: any = ["bootstrap", "ready", "serving", "draining", "stopped"];
export function inferServiceLifecycle(services: any, parser_evidence: any): any {
  var phases: any[] = [];
  var svc: any;
  for (svc of py.iter(py.sorted(services, {key: ((s: any) => py.toStr(py.get(s, "name", ""))) as (item: any) => any}))) {
    var phase: any = String(py.toStr(py.get(svc, "phase", "bootstrap"))).toLowerCase();
    if (!py.contains(LIFECYCLE_PHASES, phase)) {
      phase = "bootstrap";
    }
    py.listAppend(phases, {"name": py.get(svc, "name"), "phase": phase});
  }
  return {"services": phases, "evidence": py.sorted(py.toSet(parser_evidence)), "grounded": py.truthy(parser_evidence), "deterministic": true};
}
